{% extends "base.html" %}
{% load static i18n %}


{% block content %}
    <div class="container-fluid">
        <!-- Jumbotron -->
        <div class="card card-image" style="">
            <div class="text-center py-2 px-2">
                <div class="py-2">
            
                    <!-- Content -->
                    <h2 class="card-title h2 my-2 py-2">Technical corner</h2>

                </div>
            </div>
        </div>
        <!-- Jumbotron -->

    </div>
    <div class="container">
        <div class="row my-5">
            <div class="col-md-6">
                <h2>Study curation</h2>
                <hr>
                    <p>Each study available in RGV was manually curated to extract relevant information:</p>
                    <ul class="list-unstyled">
                        <li>Information relative to the associated publication (PubMed ID, authors, full publication name, abstract)</li>
                        <li>The species investigated and the technology used</li>
                        <li>Relevant information about investigated samples (tissues, sex, developpemental stage, ...)</li>
                        <li>Any relevant keywords</li>
                    </ul>
                    <p>If available, the raw data associated with the study will then be processed using a multiple-step workflow to make it compatible with the RGV visualization tools.</p>
            </div>
            <div class="col-md-6">
                <h2>Sample Processing</h2>
                <hr>
                <p>For each sample, seven processing steps are sequentially performed:</p>
                <ul class="list-unstyled">
                    <li>1- The raw data file (sra file) of a given sample and their associated information are downloaded from the <a href="https://www.ncbi.nlm.nih.gov/geo/" target="_blanck">Gene Expression Omnibus</a> (GEO) and to the <a href="https://www.ncbi.nlm.nih.gov/sra" target="_blanck">Sequence Read Archive</a> (SRA);</li>
                    <li>2- The fastq file is extracted from the sra archive file;</li>
                    <li>3- Reads are mapped on its corresponding reference genome using the <a href="https://www.ncbi.nlm.nih.gov/pubmed/23104886" target="_blanck">STAR</a> tool.</li>
                    <li>4- Each bam file is converted into a simple tab-delimited text file (BED) in which data are standardized according to the total nuber of read counts using <a href="https://www.ncbi.nlm.nih.gov/pubmed/20110278" target="_blanck">BEDTools</a>;</li>
                    <li>5- The standardized BED file is then converted into an indexed binary format (bigWig or bw) to enable fast remote access to the data;</li>
                    <li>6- The pairwise alignments between genome assemblies and between species provided by UCSC made it possible to convert genome coordinates in the resulting bigWig file from a genome of a given species into a genome of another species ussing <a href="https://www.ncbi.nlm.nih.gov/pubmed/24351709" target="_blanck">CrossMap</a>.</li>
                    <li>7- Reference transcripts are quantified in the corresponding sample using the <a href="https://www.ncbi.nlm.nih.gov/pubmed/25690850" target="_blanck">StringTie</a> tool.</li>
                </ul>

            </div>
        </div>

        <div class="row my-5">
            <div class="col-md-6 mb-5">
                <h2>Framework and plugins</h2>
                <hr>
                <p> The web interface of the ReproGenomics Viewer database was built with the web framework <a href="https://www.djangoproject.com/" target="_blank">Django</a>.
                <p>All charts generated by RGV's tools use the <a href="https://plot.ly/javascript/" target="_blank">plotly.js plugin</a>.</p>
                <p>The charts layouts are managed using <a href="https://golden-layout.com/" target="_blank">Golden layout</a></p>
            </div>
            <div class="col-md-6 mb-5">
                <h2>Source code</h2>
                <hr>
                    <p>The full source code of the ReproGenomics Viewer web application is available <a href="https://github.com/umr1085-irset/reproGenomicsViewer">here</a>.</p>
            </div>
        </div>        
    </div>

{% endblock %}
